qemu/tests/vsock_test.py

```diff
@@ -0,0 +1,152 @@
+    """
+    Vsock basic function test
+
+    1. Boot guest with vhost-vsock-pci device
+    2. Download and compile nc-vsock on both guest and host
+    3. Start listening inside guest, nc-vsock -l $port
+    4. Connect guest CID from host, nc-vsock $guest_cid $port
+    5. Input character, e.g. 'Hello world'
+    6. Check if guest receive the content correctly
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def clean(tmp_file=None):
+        """ Clean the environment """
+        cmd_rm = 'rm -rf %s*' % nc_vsock_bin
+        if tmp_file:
+            cmd_rm += '; rm -rf %s' % tmp_file
+        session.cmd_output_safe(cmd_rm)
+        process.system(cmd_rm, ignore_status=True)
+        if host_vsock_session.is_alive():
+            host_vsock_session.close()
+        session.close()
+
+    vm = env.get_vm(params["main_vm"])
+    tmp_file = '/tmp/vsock_file_%s' % utils_misc.generate_random_string(6)
+    session = vm.wait_for_login()
+    nc_vsock_bin = compile_nc_vsock(test, vm, session)
+    vm_ip_addr = vm.wait_for_get_address(0)
+    port = utils_misc.find_free_port(5000, 6000, vm_ip_addr)
+    vsock_dev = params["vsocks"].split()[0]
+    guest_cid = vm.devices.get(vsock_dev).get_param("guest-cid")
+    nc_vsock_listen(nc_vsock_bin, port, session)
+    host_vsock_session = nc_vsock_connect(nc_vsock_bin, guest_cid, port)
+    connected_str = r"Connection from cid*"
+    send_data = "Hello world"
+    if not session.read_until_last_line_matches(connected_str):
+        clean()
+        test.fail("Connection failed.")
+    error_context.context('Input "Hello world" to vsock.', logging.info)
+    host_vsock_session.sendline(send_data)
+    if not session.read_until_last_line_matches(send_data):
+        clean()
+        test.fail("Data transfer failed.")
+    host_vsock_session.close()
+    session.close()
+
+    # Transfer data from guest to host
+    session = vm.wait_for_login()
+    cmd_generate = 'dd if=/dev/urandom of=%s count=1000 bs=1M' % tmp_file
+    session.cmd_status(cmd_generate)
+    cmd_transfer = '%s -l %s < %s &' % (nc_vsock_bin, port, tmp_file)
+    error_context.context('Transfer file from guest via command: %s'
+                          % cmd_transfer, logging.info)
+    session.cmd_status(cmd_transfer)
+    time.sleep(1)  # wait nc-vsock start listening
+    cmd_receive = '%s %s %s > %s' % (nc_vsock_bin, guest_cid, port, tmp_file)
+    rec_session = aexpect.Expect(cmd_receive,
+                                 auto_close=True,
+                                 output_func=utils_misc.log_line,
+                                 output_params=('%s.log' % tmp_file,))
+    utils_misc.wait_for(lambda: not rec_session.is_alive(), timeout=20)
+    cmd_chksum = 'md5sum %s' % tmp_file
+    md5_origin = session.cmd_output(cmd_chksum).split()[0]
+    md5_received = process.system_output(cmd_chksum).split()[0]
```

Comment:
@vivianQizhu  In my side, test failed as md5 value are different.  Any known product bug for this test case?   Otherwise, we need to check the code logic again.

Reply:
@yanan-fu No product issue, manual test always pass, I have hit the md5 mismatch issue before previously when I use common shell session for host receiving process, but it is okay in my test after moving to `Expect`. Anyway should be code issue, I need check it. Thanks.